Made the chart fully Helm 3 compliant. Dropped support for Helm 2.

yuchaoran2011 opened the pull request:
### Type of change

- Enhancement / new feature
- Documentation

### Description

Changed directory structure to be fully Helm-3 compliant. Specifically, CRD YAMLs have been moved to a dedicated `/crds` directory. And these files cannot be templated according to Helm 3, so some labels have been removed.

### Checklist

Not sure what I should put in CHANGELOG.md. I see that 0.16 rc has been created. So it should go in the 0.17 section? Please advise. The chart has been tested on an GKE cluster with Helm 3.

yuchaoran2011 commented:
The integration test failed because this [script](https://github.com/strimzi/strimzi-kafka-operator/blob/master/.travis/setup-kubernetes.sh#L27) needs to point to a Helm 3 counterpart. Does anyone need to know where one is available?

scholzj commented:
Thanks for the PR.

Not really using and understanding Helm, I have some questions:
* If I understand it correctly, the Helm Charts for Helm 2 and Helm 3 are not compatible, right?
* Assuming it is the case, Im not sure we can really replace the Helm 2 chart since I don't think everyone migrated (for example given the new PR for the old chart from yesterday). So would it be better to add a new directory with the Helm3 Chart and keep them in parrallel for some time before the majority of people moves to Helm 3?

> The integration test failed because this script needs to point to a Helm 3 counterpart. Does anyone need to know where one is available?

I guess we would need to write our own if there is not one available. We pretty much just need to download the binarry I guess.

yuchaoran2011 commented:
Thanks @scholzj for the comment. To answer your questions,
* Helm 2 charts are theoretically installable using Helm 3 binary, although minor changes might be required. For charts that are fully Helm-3 compliant, Helm 2 binary won't be able to install them. I think what you suggested is a good idea. I'll create another directory for the Helm 3 chart.
* I found a working script that we can use: https://github.com/helm/helm/blob/master/scripts/get-helm-3. It's in master and not in any of the existing release branches yet.

yuchaoran2011 commented:
I have moved the Helm 3 chart to a separate directory and updated Travis to use Helm 3 instead. The build still failed with an error:
```
ERROR: Uncommitted changes in derived resources:
M	helm-charts/strimzi-kafka-operator/templates/047-Crd-kafkaconnector.yaml
```
I wonder what it means?

dalelane commented on .travis.yml:
```diff
@@ -32,7 +32,7 @@ env:
   - TEST_KUBECTL_VERSION=v1.15.0
   - TEST_NSENTER_VERSION=2.32
-  - TEST_HELM_VERSION=v2.9.1
+  - TEST_HELM_VERSION=v3.0.2
```
While the Helm 2 charts are still being kept, I think we should still keep the tests for them, so it'd be better to introduce a new env var for testing Helm 3 (and avoid any possible breaking changes by relying on testing Helm 2 charts using the v3 CLI).

dalelane commented on helm-charts/strimzi-kafka-operator-helm3/Chart.yaml:
```diff
@@ -0,0 +1,21 @@
+apiVersion: v2
+appVersion: "0.1.0"
+description: "Strimzi: Apache Kafka running on Kubernetes"
+name: strimzi-kafka-operator-helm3
```
You could restructure the folder layout to avoid needing to rename the chart. That might be nicer than building something with "helm3" in the name.

e.g. `helm-charts/strimzi-kafka-operator/Chart.yaml` could be the existing one, and you could add
`helm-3-charts/strimzi-kafka-operator/Chart.yaml` for the new one, which can still be called strimzi-kafka-operator

That way each folder could have it's own Makefile, one built and linted with Helm 2, the other built and linted with Helm 3.

Or, if you want to keep them all in a single folder with a single Makefile, you could do something like:
`helm-charts/helm-2/strimzi-kafka-operator/Chart.yaml` for the existing one, and
`helm-charts/helm-3/strimzi-kafka-operator/Chart.yaml` for the new one

dalelane commented on helm-charts/strimzi-kafka-operator-helm3/README.md:
````diff
@@ -0,0 +1,112 @@
+To install the chart with the release name `my-release`:
+
+```bash
+$ helm install --name my-release strimzi/strimzi-kafka-operator
````
Unless you change the chart name (as mentioned above) this will need updating.

dalelane commented on helm-charts/strimzi-kafka-operator-helm3/crds/040-Crd-kafka.yaml:
```diff
@@ -0,0 +1,3686 @@
+    openAPIV3Schema:
+      properties:
+        spec:
+          type: object
```
In the README, you've said a minimum Kubernetes version of 1.9+. 

In https://github.com/strimzi/strimzi-kafka-operator/pull/2232#issuecomment-557465149 @tombentley was suggesting that having this top-level `type: object` value is only supported from 1.12 onwards.

dalelane commented on helm-charts/strimzi-kafka-operator-helm3/crds/040-Crd-kafka.yaml:
```diff
@@ -0,0 +1,3686 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
How will these be kept updated along the existing Helm 2 charts? Are you planning to make any updates to the Makefile to generate/update these in future? (Or write a new Makefile for building the Helm 3 charts?)

dalelane commented on helm-charts/strimzi-kafka-operator/README.md:
```diff
@@ -45,8 +42,7 @@ The command removes all the Kubernetes components associated with the operator a
 The following table lists the configurable parameters of the Strimzi chart and their default values.  Runtime
-configuration of Kafka and other components are defined within their respective Custom Resource Definitions.  See
-the documentation for more details.
+configuration of Kafka and other components are defined within their respective Custom Resource Definitions. See the documentation for more details.
```
Is the newline needed here?

scholzj commented:
@yuchaoran2011 any plans to get back to this and look at the comments? Or should someone else try to continue on this? Thanks.

yuchaoran2011 commented:
@scholzj Sorry I got busy at work and won't have time to finish this PR anytime soon. Feel free to have someone else take it over.

scholzj commented:
@yuchaoran2011 Thanks. I opened an issue #2895 linking back here so that we can find it when we get back to it and resue the work from here.
